### PR TITLE
Added Fernet encrypter.

### PR DESCRIPTION
**Pull request by rohe**

### CODE REVIEW

File: src/cryptojwt/jwe/fernet.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class FernetEncrypter(Encrypter):
+    def __init__(self, password: str, salt: Optional[bytes] = ""):
+        Encrypter.__init__(self)
+        if not salt:
+            salt = os.urandom(16)
+        else:
+            salt = as_bytes(salt)
+
+        kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=390000)
```
Comment (peppelinux) [inline src/cryptojwt/jwe/fernet.py]: may we get a str here, like "SHA256" and then try a getattr(hashlib, algorithm)?

Comment (rohe) [reply src/cryptojwt/jwe/fernet.py]: OK

File: pyproject.toml [hunk 2]
```diff
@@ -22,7 +22,7 @@ exclude_lines = [
 
 [tool.poetry]
 name = "cryptojwt"
-version = "1.7.1"
+version = "1.7.2"
```
Comment (jschlyter) [inline pyproject.toml]: This is an API change/update, should be bumped to 1.8.0

File: src/cryptojwt/jwe/fernet.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+
+from cryptojwt import as_unicode
+from cryptojwt.jwe import Encrypter
+from cryptojwt.utils import as_bytes
+
+
+class FernetEncrypter(Encrypter):
+    def __init__(self, password: str, salt: Optional[bytes] = ""):
```
Comment (jschlyter) [inline src/cryptojwt/jwe/fernet.py]: `= None`  if optional?

File: src/cryptojwt/jwe/fernet.py (hunk 1 above)
Comment (jschlyter) [reply src/cryptojwt/jwe/fernet.py]: Also, use CONSTANT for 390000, e.g. `DEFAULT_PBKDF2HMAC_ITERATIONS`  or so

**Review by jschlyter — DISMISSED** [review]

File: pyproject.toml (hunk 2 above)
Comment (rohe) [reply pyproject.toml]: OK

File: src/cryptojwt/jwe/fernet.py (hunk 3 above)
Comment (rohe) [reply src/cryptojwt/jwe/fernet.py]: Not sure why None would be better than "" or b"" for that matter ?

File: src/cryptojwt/jwe/fernet.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+DEFAULT_ITERATIONS = 390000
+
+
+class FernetEncrypter(Encrypter):
+    def __init__(
+        self,
+        password: str,
+        salt: Optional[bytes] = "",
+        hash_alg: Optional[str] = "SHA256",
+        digest_size: Optional[int] = 0,
```
Comment (jschlyter) [inline src/cryptojwt/jwe/fernet.py:22]: should be `= None` ?

File: src/cryptojwt/jwe/fernet.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        Encrypter.__init__(self)
+        if not salt:
+            salt = os.urandom(16)
+        else:
+            salt = as_bytes(salt)
+
+        _alg = getattr(hashes, hash_alg)
+        # A bit special for SHAKE* and BLAKE* hashes
+        if hash_alg.startswith("SHAKE") or hash_alg.startswith("BLAKE"):
+            _algorithm = _alg(digest_size)
```
Comment (jschlyter) [inline src/cryptojwt/jwe/fernet.py:34]: What if digest size is 0 or None?

Perhaps `digest_size or DEFAULT_DIGEST_SIZE` and define the a default?

**Review by jschlyter — APPROVED** [review]
Looking good, perhaps fix default digest size

File: src/cryptojwt/jwe/fernet.py (hunk 5 above)
Comment (rohe) [reply src/cryptojwt/jwe/fernet.py:34]: There doesn't seem to be a simple default size.
0 is allowed by the SHAKE* algorithms.
The two BLAKE* algorithms have different digest_sizes.
Hence, no easy default.

File: src/cryptojwt/jwe/fernet.py (hunk 3 above)
Comment (jschlyter) [reply src/cryptojwt/jwe/fernet.py]: If you say `Optional`, it makes sense to make the default  `None` to indicate that no value (not even a zero length string) as provided. Mostly a question of style I guess.

File: src/cryptojwt/jwe/fernet.py (hunk 5 above)
Comment (jschlyter) [reply src/cryptojwt/jwe/fernet.py:34]: If the `digest_size` is required, we should raise a `ValueError` if not provided (when required).

Comment (rohe) [reply src/cryptojwt/jwe/fernet.py:34]: I think that's up to the underlying library.
Things may change and I don't want to have to track that.

File: src/cryptojwt/jwe/fernet.py (hunk 3 above)
Comment (rohe) [reply src/cryptojwt/jwe/fernet.py]: I guess :-)

**Comment by jschlyter** [comment]
I'm good.

**Pull request merged**